presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/Tuple.java

```diff
@@ -0,0 +1,144 @@
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class Tuple
+        implements Comparable<Tuple>
+{
+    private final List<Type> types;
+    private final List<Object> values;
+
+    public Tuple(Type type, Object value)
+    {
+        this(ImmutableList.of(type), ImmutableList.of(value));
+    }
+
+    public Tuple(List<Type> types, Object... values)
+    {
+        this(types, ImmutableList.copyOf(values));
+    }
+
+    public Tuple(List<Type> types, List<Object> values)
+    {
+        this.types = requireNonNull(types, "types is null");
+        this.values = requireNonNull(values, "values is null");
+
+        checkArgument(!types.isEmpty(), "types is empty");
+        checkArgument(types.size() == values.size(), "types and values must have the same number of elements");
+    }
+
+    public List<Type> getTypes()
+    {
+        return types;
+    }
+
+    public List<Object> getValues()
+    {
+        return values;
+    }
+
+    @Override
+    public int compareTo(Tuple o)
+    {
+        checkArgument(o.getTypes().size() == this.types.size(), "types must be of same size");
+        for (int i = 0; i < types.size(); i++) {
+            checkArgument(o.getTypes().get(i) == this.types.get(i), "types must be the same");
+        }
+
+        for (int i = 0; i < types.size(); i++) {
+            Type type = types.get(i);
+            Object o1 = values.get(i);
+            Object o2 = o.getValues().get(i);
+
+            int result = compare(type, o1, o2);
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        Tuple tuple = (Tuple) o;
+        return Objects.equals(types, tuple.types) &&
+                Objects.equals(values, tuple.values);
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(types, values);
+    }
+
+    private static int compare(Type type, Object o1, Object o2)
+    {
+        if (type.equals(BOOLEAN)) {
+            return Boolean.compare((Boolean) o1, (Boolean) o2);
+        }
+        if (type.equals(INTEGER)) {
+            return Integer.compare((Integer) o1, (Integer) o2);
+        }
+        if (type.equals(BIGINT)) {
+            return Long.compare((Long) o1, (Long) o2);
+        }
+        if (type.equals(DOUBLE)) {
+            return Doubles.compare((Double) o1, (Double) o2);
+        }
+        if (type.equals(VARCHAR)) {
```

Comment:
This won't work for parameterized varchar with different lengths. For example, `varchar(10)` and `varchar(20)` are not equal. Is that a problem here?